cpp/src/groupby/sort/groupby.cu

```diff
@@ -14,32 +14,293 @@
-std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> groupby(
-    table_view const& keys, std::vector<aggregation_request> const& requests,
-    cudaStream_t stream, rmm::mr::device_memory_resource* mr) {
-  // stub
-  return std::make_pair(std::make_unique<table>(),
-                        std::vector<aggregation_result>{});
+
+std::vector<aggregation_result> extract_results(
+    std::vector<aggregation_request> const& requests,
+    result_cache& cache)
+{
+  std::vector<aggregation_result> results(requests.size());
+
+  for (size_t i = 0; i < requests.size(); i++) {
+    for (auto &&agg : requests[i].aggregations) {
+      results[i].results.emplace_back( cache.release_result(i, agg) );      
+    }
+  }
+  return results;
 }
-}  // namespace sort
+
 }  // namespace detail
+
+// Sort-based groupby
+std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> 
+groupby::sort_aggregate(
+    std::vector<aggregation_request> const& requests,
+    cudaStream_t stream, rmm::mr::device_memory_resource* mr)
+{
+  // We're going to start by creating a cache of results so that aggs that
+  // depend on other aggs will not have to be recalculated. e.g. mean depends on
+  // sum and count. std depends on mean and count
+  detail::result_cache cache(requests.size());
+  
+  for (size_t i = 0; i < requests.size(); i++) {
+    // TODO (dm): Use key_sorted_order to make permutation iterator and avoid
+    //            generating value columns
+    std::unique_ptr<column> sorted_values;
+    std::unique_ptr<column> grouped_values;
+
+    auto get_grouped_values = 
+      [i, this, &grouped_values, &sorted_values, &requests] () 
+    {
+      // TODO (dm): After implementing single pass mutli-agg, explore making a
+      //            cache of all grouped value columns rather than one at a time
+      if (grouped_values)
+        return grouped_values->view();
+      else if (sorted_values)
+        // TODO (dm): When we implement scan, it wouldn't be ok to return sorted
+        //            values when asked for grouped values. Change this then.
+        return sorted_values->view();
+      else
+        grouped_values = this->helper().grouped_values(requests[i].values);
+      return grouped_values->view();
+    };
+
+    auto get_sorted_values = 
+      [i, this, &sorted_values, &requests] () 
+    {
+      if (not sorted_values)
+        sorted_values = this->helper().sorted_values(requests[i].values);
+      return sorted_values->view();
+    };
+
+    auto store_count = 
+      [i, &cache, &stream, &mr, this, get_grouped_values] 
+      (std::unique_ptr<aggregation> const& agg)
+    {
+      if (cache.has_result(i, agg))
+        return;
+
+      cache.add_result(i, agg, 
+                       detail::group_count(get_grouped_values(), 
+                                this->helper().group_labels(),
+                                this->helper().num_groups(), mr, stream));
+    };
+
+    auto store_sum =
```

Comment:
This style of putting everything in lambdas makes it hard to understand what the inputs/outputs and potential side effects of each of these are. It'd be easier to follow if these were free functions.

Reply:
Or potentially a single function with a non-type template for `aggregation::Kind` and then the aggregation dispatcher can be used.

Reply:
This made more sense when I was default capturing everything by reference. Now that I have to list out all the captures, I think I could've just used free functions. Plus I just added dependency on the new aggregation dispatcher that takes args.